### ISSUE TITLE
CMake errors in first step of build process (cd build && cmake ..)

### ISSUE THREAD

**Issue by robert-dodier** [wxMaxima-developers/wxmaxima#1388]
I'm trying to follow the instructions in Compiling.md and running into errors at the first step. I tried `mkdir build && cd build && cmake ..` and at this point I get the errors shown below. It appears that CMake is not reading CMakeLists.txt correctly, because that file does contain `project` and it does contain a definition for `list_transform_prepend`. What else can I try?

I have build wxMaxima before, but it has been months or years, and I don't remember how it worked before. For the record I am working with Xubuntu 14.04 with cmake 3.16.4. 

PS. Here is the cmake output:
```
wxmaxima/build$ cmake ..
CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as

    project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at CMakeLists.txt:32 (list_transform_prepend):
  Unknown CMake command "list_transform_prepend".


CMake Warning (dev) in CMakeLists.txt:
  No cmake_minimum_required command is present.  A line of code such as

    cmake_minimum_required(VERSION 3.16)

  should be added at the top of the file.  The version specified may be lower
  if you wish to support older CMake versions for this project.  For more
  information run "cmake --help-policy CMP0000".
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring incomplete, errors occurred!
See also "/home/robert/github/github-forks/wxmaxima/CMakeFiles/CMakeOutput.log".
```

**Comment by gunterkoenigsmann** [comment]
Wow... ...I am using cmake 3.16.3 and it works fine => currently trying to upgrade my cmake in order to find out what happened.

---

If you are working on data/wxmathML.lisp: In order to get your changes compiled into wxMaxima (and therefore get them used by wxMaxima) data/wxmathML.lisp first needs to be converted into a .h file - which is done by running `./generate_wxmathml.sh` in the directory `data.

**Comment by gunterkoenigsmann** [comment]
Updated cmake to cmake 3.17.4 and it still is able to read CMakeLists.txt.

A typical problem with many programs is that the path to the build directory contains spaces and/or unicode letters. But I think you would have thought about that...

**Comment by gunterkoenigsmann** [comment]
Hmmm... ...the only thing stackoverflow tells me for this error message is that the build directory might be in the src/ dir, not in the top-level dir of wxMaxima (one level above src/)

**Comment by gunterkoenigsmann** [comment]
Ok... ...the top-level CMakeLists.txt now contains a literal project() command that is not hidden inside an `if()`. If that makes all cmake versions detect it?

**Comment by KubaO** [comment]
I believe that this issue has been resolved. Please reopen if that's not the case for you.

**Comment by robert-dodier** [comment]
I've tried it again, this time on macOS Catalina 10.15.7, with cmake 3.19.2, trying to compile current version from Git (commit a473cb). Same error from `cmake -S . -B ../build-wxm`:

CMake Error at CMakeLists.txt:33 (list_transform_prepend):
  Unknown CMake command "list_transform_prepend".

**Comment by gunterkoenigsmann** [comment]
Seems like you are a more valuable tester than you want to as your system seems to be able to reproduce many kinds of problems the CI doesn't find. With a bit of luck commit 07996d9 resolves the new issue.

**Comment by robert-dodier** [comment]
OK, after updating from Git, I get:
```{text}
CMake Error at /usr/local/Cellar/cmake/3.19.2/share/cmake/Modules/CheckIPOSupported.cmake:147 (message):
  Policy CMP0069 is not set
Call Stack (most recent call first):
  CMakeLists.txt:160 (check_ipo_supported)
```
Dunno what that's about.

**Comment by gunterkoenigsmann** [comment]
If cmake introduces possibly-breaking changes they document that change as a "policy change". 
If your CMakeLists.txt claims that you support a pre-policy-change cmake they don't dare to apply the change to your project and issue a warning about that. Together that fells like they know what they doing. But it seems that this time in at least one CMake version they (accidentally?) issued an error, instead while other cmake versions issue a warning, instead => tried to resolve this bug. If it still appears or you happen to trigger another bug I am happy to try to fix any additional bug. If my last commit unbreaks things for you, instead, please tell me: My policy is to make a new release once I have been able to un-break a platform.

**Comment by robert-dodier** [comment]
After updating from Git, I get the same error. I notice that there was already code in CMakeLists.txt to set CMP0069, so whatever it was, it wasn't working before either.

**Comment by gunterkoenigsmann** [comment]
Ok... ...with a bit of luck compilation now works with the git master.

**Comment by robert-dodier** [comment]
Gunter, thanks for the update. I tried it again and I still see the same error. 

I notice that `cmake --help-policy CMP0069` prints some advice about how to deal with IPO. Maybe that can be adapted in some way.

I am working with cmake 3.19.2 because that is the version that was installed by Homebrew. If you are working with a different version, and the configuration files have an implicit dependence on that version, maybe the cmake version can be assigned. It looks like there is a way to make cmake act like an older version, although I'm not sure about how to do that.

**Comment by gunterkoenigsmann** [comment]
> I am working with cmake 3.19.2 because that is the version that was installed by Homebrew. If you are working with a different version, and the configuration files have an implicit dependence on that version, maybe the cmake version can be assigned. It looks like there is a way to make cmake act like an older version, although I'm not sure about how to do that.

That's one of the reasons why I am happy if I can use you as a tester. The absurd thing is that the build runs just fine with cmake 3.19.1 on the mac => Gave it a last try. If that doesn't work perhaps it is wise to drop support for link-time optimizations on the mac for now...

**Comment by robert-dodier** [comment]
OK, I've tried it again with the current version, and I get an error about Info.plist being missing. I see there's Info.plist.in somewhere. In a build system based on autotools, I would expect Info.plist to be automatically generated from Info.plist.in. Is cmake supposed to execute that step? There are other something.in files so I would expect that they all need to be processed. Is there another way to generate the something files from something.in?

**Comment by gunterkoenigsmann** [comment]
Sorry that this took longer than usual. But according to https://github.com/wxMaxima-developers/wxmaxima/runs/1709045674?check_suite_focus=false the Info.plist should now be placed inside the bundle in the `make install` phase.

**Comment by gunterkoenigsmann** [comment]
The Info.plist file should end up in the data folder of the build directory. I have now told the `CMakelists.txt` to generate this file even if cmake isn't sure that this actually is a mac.

**Comment by robert-dodier** [comment]
I've updated to commit 054598 and I'm still getting the error about Info.plist. I see that Info.plist doesn't exist anywhere, so I guess it isn't getting generated as expected. I'm executing `cmake -B ../build-wxm -S .` (I don't have Ninja installed and an attempt to install it was unsuccessful so I am trying to run without it). Is there some other stuff that I should execute before that, in order to get Info.plist generated? Thanks for your help, I appreciate it.

**Comment by gunterkoenigsmann** [comment]
And I am very happy to have someone who knows how to debug problems, owns a mac *and* is willing to help...

...but this time I am afraid to run out of ideas before the problem is solved: Theoretically in the current git master when setting up the makefiles `data/CMakeLists.txt` should unconditionally generate a `data/Info.plist` in the build dir out of `data/Info.plist.in` in the source dir. ...and normally it should say that it does this in the first `cmake` step of the build instructions.

Could you send me your build log? And the log from the `cmake --install ` step? Perhaps I can see there what goes wrong...

**Comment by robert-dodier** [comment]
I've attached build-wxm/CMakeFiles/CMakeOutput.log:
[CMakeOutput.log](https://github.com/wxMaxima-developers/wxmaxima/files/5824937/CMakeOutput.log)

I don't get any farther than `cd src && cmake -B ../build-wxm -S .`, so there is nothing from `cmake --install` to show.

**Comment by gunterkoenigsmann** [comment]
I don't own a mac so I normally don't dare to change anything mac-specific. What I can do is to tell or more mac computers to compile every single build we make so I tried to tell such a computer to use the official build instructions - which surprisingly for me didn't work.

What seems to work, at least with a fresh checkout of wxMaxima's master branch, is:

           mkdir ../build-wxm
           cmake -S . -B ../build-wxm -DCMAKE_INSTALL_PREFIX=.
           ls -l ../build-wxm/data/Info.plist
           cmake --build ../build-wxm
           cmake --install ../build-wxm

https://github.com/wxMaxima-developers/wxmaxima/actions/runs/491521623 contains the logs of an compiler run with ninja as build tool (ninja is *fast*) and one compiler run with make as build tool.
As the `ls` shows (or at least should show in your case) the plist file should be generated in the first of these steps.

One last thing I don't understand: You told you have tried commit 054598 - and I don't see such a commit => Might there be something wrong with your local git repo?

**Comment by robert-dodier** [comment]
I've updated to 9efa96 and I get the same error about Info.plist from `cd src && cmake -B ../build-wxm -S .`. 

One thing I have discovered is that `configure_file(.../Info.plist.in, .../Info.plist)` is specified in `data/CMakeLists.txt` and not in `src/CMakeLists.txt`. I find that `cd data && cmake -B ../build-wxm -S .` generates `build-wxm/data/Info.plist` as expected. However, then when I do `cd ../src && cmake -B ../build-wxm -S .`, I get an error:

CMake Error: The source "/Users/dodier/github/wxmaxima/src/CMakeLists.txt" does not match the source "/Users/dodier/github/wxmaxima/data/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.

At this point I have to give up. I'll close this item, but if at some point in the future, someone figures out what to do, I'll be interested to hear about it.
